File: tests/security-capability/semgrep/go/owned_core.go
package fixture

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	crand "crypto/rand"
	"database/sql"
	"encoding/gob"
	"fmt"
	"html/template"
	"math/rand"
	"net/http"
	"os"
	"os/exec"
)

func vulnerable(input string, db *sql.DB, decoder *gob.Decoder, target any) {
	// ruleid: vuln-workbench.go.command-injection
	exec.Command(input)
	// ruleid: vuln-workbench.go.command-injection
	exec.Command(input, "--version")
	// ruleid: vuln-workbench.go.sql-injection
	db.Query(fmt.Sprintf("SELECT * FROM users WHERE id = %s", input))
	// ruleid: vuln-workbench.go.sql-injection
	db.Exec(fmt.Sprintf("DELETE FROM users WHERE id = %s", input))
	// ruleid: vuln-workbench.go.template-html
	template.HTML(input)
	// ruleid: vuln-workbench.go.template-html
	_ = template.HTML(input)
	// ruleid: vuln-workbench.go.ssrf-http-get
	http.Get(input)
	// ruleid: vuln-workbench.go.ssrf-http-get
	http.Get(input + "/health")
	// ruleid: vuln-workbench.go.path-traversal-open
	os.Open(input)
	// ruleid: vuln-workbench.go.path-traversal-open
	os.Open(input + ".json")
	// ruleid: vuln-workbench.go.unsafe-gob-decode
	gob.NewDecoder(os.Stdin).Decode(target)
	// ruleid: vuln-workbench.go.unsafe-gob-decode
	gob.NewDecoder(os.Stdin).Decode(&target)
	// ruleid: vuln-workbench.go.weak-hash
	md5.New()
	// ruleid: vuln-workbench.go.weak-hash
	sha1.New()
	// ruleid: vuln-workbench.go.weak-random
	rand.Intn(10)
	// ruleid: vuln-workbench.go.weak-random
	rand.Read([]byte{0})
}

func fixed(db *sql.DB) {
	// ok: vuln-workbench.go.command-injection
	exec.Command("/usr/bin/id", "-u")
	// ok: vuln-workbench.go.command-injection
	exec.Command("/usr/bin/date")
	// ok: vuln-workbench.go.sql-injection
	db.Query("SELECT * FROM users WHERE id = ?", 1)
	// ok: vuln-workbench.go.sql-injection
	db.Exec("DELETE FROM users WHERE id = ?", 1)
	// ok: vuln-workbench.go.template-html
	template.HTMLEscapeString("safe")
	// ok: vuln-workbench.go.template-html
	template.JSEscapeString("safe")
	// ok: vuln-workbench.go.ssrf-http-get
	http.Get("https://example.invalid/health")
	// ok: vuln-workbench.go.ssrf-http-get
	http.Get("/health")
	// ok: vuln-workbench.go.path-traversal-open
	os.Open("settings.json")
	// ok: vuln-workbench.go.path-traversal-open
	os.Open("/srv/static/index.html")
	// ok: vuln-workbench.go.unsafe-gob-decode
	fmt.Println("safe")
	// ok: vuln-workbench.go.unsafe-gob-decode
	fmt.Sprintf("%s", "safe")
	// ok: vuln-workbench.go.weak-hash
	sha256.New()
	// ok: vuln-workbench.go.weak-hash
	sha256.Sum256([]byte("safe"))
	// ok: vuln-workbench.go.weak-random
	crand.Read([]byte{0})
	// ok: vuln-workbench.go.weak-random
	crand.Reader.Read([]byte{0})
}

File: tests/security-capability/semgrep/typescript/owned-core.ts
import child_process from "node:child_process";
import crypto from "node:crypto";
import type { Response } from "express";
import axios from "axios";
import jwt from "jsonwebtoken";

declare const db: { query: (sql: string) => unknown };

export function vulnerable(
	input: string,
	element: HTMLElement,
	response: Response,
) {
	// ruleid: vuln-workbench.typescript.command-injection
	child_process.exec(input);
	// ruleid: vuln-workbench.typescript.command-injection
	child_process.execSync(input);
	// ruleid: vuln-workbench.typescript.sql-injection
	db.query(`SELECT * FROM users WHERE id = ${input}`);
	// ruleid: vuln-workbench.typescript.sql-injection
	db.query(`DELETE FROM users WHERE id = ${input}`);
	// ruleid: vuln-workbench.typescript.dom-xss
	element.innerHTML = input;
	// ruleid: vuln-workbench.typescript.dom-xss
	document.body.innerHTML = input;
	// ruleid: vuln-workbench.typescript.ssrf-axios
	axios.get(input);
	// ruleid: vuln-workbench.typescript.ssrf-axios
	axios.post(input);
	// ruleid: vuln-workbench.typescript.path-traversal-sendfile
	response.sendFile(input);
	// ruleid: vuln-workbench.typescript.path-traversal-sendfile
	response.sendFile(input, {});
	// ruleid: vuln-workbench.typescript.weak-hash
	crypto.createHash("md5");
	// ruleid: vuln-workbench.typescript.weak-hash
	crypto.createHash("sha1");
	// ruleid: vuln-workbench.typescript.jwt-decode-auth
	jwt.decode(input);
	// ruleid: vuln-workbench.typescript.jwt-decode-auth
	return jwt.decode(input, { json: true });
}

export function logSecrets(password: string, apiKey: string) {
	// ruleid: vuln-workbench.typescript.sensitive-console-log
	console.log(password);
	// ruleid: vuln-workbench.typescript.sensitive-console-log
	console.log("key", apiKey);
}

export function fixed(response: Response, element: HTMLElement) {
	// ok: vuln-workbench.typescript.command-injection
	child_process.execFile("/usr/bin/id", ["-u"]);
	// ok: vuln-workbench.typescript.command-injection
	child_process.exec("date");
	// ok: vuln-workbench.typescript.sql-injection
	db.query("SELECT 1");
	// ok: vuln-workbench.typescript.sql-injection
	db.query("SELECT * FROM users WHERE id = ?");
	// ok: vuln-workbench.typescript.dom-xss
	element.textContent = "safe";
	// ok: vuln-workbench.typescript.dom-xss
	element.innerHTML = "<b>safe</b>";
	// ok: vuln-workbench.typescript.ssrf-axios
	axios.get("https://example.invalid/health");
	// ok: vuln-workbench.typescript.ssrf-axios
	axios.post("/api/jobs");
	// ok: vuln-workbench.typescript.path-traversal-sendfile
	response.sendFile("index.html");
	// ok: vuln-workbench.typescript.path-traversal-sendfile
	response.sendFile("health.txt", { root: "/srv/static" });
	// ok: vuln-workbench.typescript.weak-hash
	crypto.createHash("sha256");
	// ok: vuln-workbench.typescript.weak-hash
	crypto.createHash("sha512");
	// ok: vuln-workbench.typescript.jwt-decode-auth
	jwt.verify("token", "public-key");
	// ok: vuln-workbench.typescript.jwt-decode-auth
	jwt.verify("token", "public-key", { algorithms: ["RS256"] });
	// ok: vuln-workbench.typescript.sensitive-console-log
	console.log("request complete");
	// ok: vuln-workbench.typescript.sensitive-console-log
	console.log("requestId", "opaque");
}
